task.go

```diff
@@ -34,28 +36,77 @@ var (
 		http.StatusTemporaryRedirect: true, // 307
 		http.StatusPermanentRedirect: true, // 308
 	}
+
+	errReferenceLoop = errors.New("detected cyclic reference")
 )
 
-type packagerTask struct {
+type packagerTaskRunner struct {
 	*Packager
+
+	period *exchange.ValidPeriod
+	errs   *multierror.MultiError
+	active map[string]bool // Keyed by URLs.
```

Comment:
It's totally up to you, but personally speaking I prefer `map[string]struct{}` to implement Set in golang. I agree that `map[string]bool` is more idiomatic but I saw only 3 lines accessing `active` and I doubt that idiomaticity makes a big difference here.

Reply:
I wasn't aware of `map[string]struct{}` so it was a good learning, but I'd stick on [Effective Go](https://golang.org/doc/effective_go.html#maps) until I find `map[string]struct{}` popular than `map[string]bool`. I don't care extra four(?) bytes per entry because keys (URL strings) consume much larger memory space. Also, in general, I don't like gaining performance merit in exchange to code readability.